File: tools/vqx/tests/test-vqx.mjs
#!/usr/bin/env node
/** VQX 0.3 JS codec tests. */
import { readFileSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { bytesToPua,puaToBytes,encodeNames,decodeIds,parseMessage,stripBootstrap,LocalTable,wrapBootstrap,BEACON,verifySha256 } from "../codecs/vqx.mjs";
const here=dirname(fileURLToPath(import.meta.url));
function findLexicon(){const env=process.env.VQX_ROOT;const c=[];if(env)c.push(join(env,"lexicon.json"),join(env,"machine","lexicon.json"));c.push(join(here,"..","lexicon.json"),join(here,"..","..","..","site","public","sites","vqx","machine","lexicon.json"));for(const p of c)if(existsSync(p))return {lex:JSON.parse(readFileSync(p,"utf8")),raw:readFileSync(p)};throw new Error("lexicon.json not found");}
const {lex,raw:lexRaw}=findLexicon();
function eq(a,b,m="assert"){const sa=JSON.stringify(a),sb=JSON.stringify(b);if(sa!==sb)throw new Error(`${m}: ${sa} != ${sb}`)}
function expect(code,fn){let got=null;try{fn()}catch(e){got=e.code}if(got!==code)throw new Error(`expected ${code}, got ${got}`)}
for(let i=0;i<256;i++){const b=Uint8Array.from([i]),p=bytesToPua(b);if(puaToBytes(p)[0]!==i||p.codePointAt(0)!==0xe000+i)throw new Error("roundtrip "+i)}
const raw=Uint8Array.from([6,17,32,160]);eq(decodeIds(raw,lex),["REQUEST","PEER","RESPOND","GLYPH_ONLY"]);eq(Array.from(encodeNames(["REQUEST","PEER","RESPOND","GLYPH_ONLY"],lex)),[6,17,32,160]);
const sample=Uint8Array.from([0xd3,0xa7,0x5c,0xe1,0x9b,0x02,0x03,0,6,17,32,160]);const parsed=await parseMessage(sample,lex,{mode:"bootstrap"});eq(parsed.version,3);eq(parsed.names,["REQUEST","PEER","RESPOND","GLYPH_ONLY"]);
expect("VQX_VERSION",()=>stripBootstrap(Uint8Array.from([...BEACON,0x99,0,6]),{mode:"bootstrap"}));expect("VQX_FLAGS",()=>stripBootstrap(Uint8Array.from([...BEACON,3,1,6]),{mode:"bootstrap"}));expect("VQX_BEACON",()=>stripBootstrap(Uint8Array.from([6]),{mode:"bootstrap"}));
const collision=Uint8Array.from([...BEACON,6]);eq(stripBootstrap(collision,{mode:"compact"}).mode,"compact");
await verifySha256(lexRaw, await (async()=>{const {createHash}=await import("node:crypto");return createHash("sha256").update(lexRaw).digest("hex")})());
let digestFailed=false;try{await verifySha256(lexRaw,"00".repeat(32))}catch(e){digestFailed=e.code==="VQX_DIGEST"}if(!digestFailed)throw new Error("digest fail-closed");
const graph=encodeNames(["REQUEST","PEER","RESPOND","GLYPH_ONLY","KEEP_CONTEXT","KEEP_CONSTRAINTS","SEARCH_IF","TOOL_IF","ACCURACY_FIRST","FINAL_ONLY"],lex);const framed=Uint8Array.from([0xdc,0xe0,graph.length,...graph,0xdd,0xe0,0xdd,0xe0,0xdd,0xe0]);const table=new LocalTable();eq(Array.from(table.expand(framed)),[...graph,...graph,...graph]);expect("VQX_TRUNCATED_MACRO",()=>new LocalTable().expand(Uint8Array.from([0xdc])));expect("VQX_TRUNCATED_MACRO",()=>new LocalTable().expand(Uint8Array.from([0xdd])));expect("VQX_LOCAL_UNDEF",()=>new LocalTable().expand(Uint8Array.from([0xdd,0xe0])));expect("VQX_LOCAL_DIRECT",()=>new LocalTable().expand(Uint8Array.from([0xe0])));expect("VQX_LOCAL_CONTENT",()=>new LocalTable().define(0xe0,Uint8Array.from([0xdd])));expect("VQX_EXPANSION_LIMIT",()=>new LocalTable().expand(framed,{maxExpandedSize:4}));
const auth=await parseMessage(encodeNames(["REQUEST","PEER","EXECUTE","TOOL"],lex),lex,{mode:"compact"});eq(auth.authorizationRequired,["EXECUTE"]);expect("VQX_FRAME_SIZE",()=>wrapBootstrap(Uint8Array.from([1,2,3,4,5]),{maxFrameSize:4}));expect("VQX_NON_PUA",()=>puaToBytes("hello",{strict:true}));
const hashed=await parseMessage(encodeNames(["ACK"],lex),lex,{mode:"compact",lexiconBytes:lexRaw,expectedDictHash:await (async()=>{const {createHash}=await import("node:crypto");return createHash("sha256").update(lexRaw).digest("hex")})()});
eq(hashed.names,["ACK"]);
let dictFail=false;try{await parseMessage(encodeNames(["ACK"],lex),lex,{mode:"compact",lexiconBytes:lexRaw,expectedDictHash:"00".repeat(32)})}catch(e){dictFail=e.code==="VQX_DIGEST"}if(!dictFail)throw new Error("parseMessage digest fail-closed");
console.log("js codec tests ok");

File: tools/vqx/tests/test_vqx.py
"""VQX 0.3 Python codec tests."""
from __future__ import annotations

import hashlib
import json
import os
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
for cand in (HERE.parent / "codecs", HERE.parent):
    if (cand / "vqx.py").exists() and str(cand) not in sys.path:
        sys.path.insert(0, str(cand)); break
import vqx  # noqa: E402


def load_lexicon():
    env = os.environ.get("VQX_ROOT")
    candidates = []
    if env:
        root = Path(env); candidates += [root / "lexicon.json", root / "machine" / "lexicon.json"]
    candidates += [HERE.parent / "lexicon.json", HERE.parent.parent.parent / "site" / "public" / "sites" / "vqx" / "machine" / "lexicon.json", HERE.parent / "machine" / "lexicon.json"]
    for p in candidates:
        if p.exists():
            raw = p.read_bytes(); return json.loads(raw), raw, p
    raise SystemExit("lexicon.json not found")


def expect(code, fn):
    try: fn()
    except vqx.VqxError as e:
        assert e.code == code, (e.code, code); return
    raise AssertionError(f"expected {code}")


def test_roundtrip_pua():
    for i in range(256):
        b=bytes([i]); p=vqx.bytes_to_pua(b)
        assert vqx.pua_to_bytes(p)==b and ord(p)==0xE000+i


def test_legacy():
    lex,_,_=load_lexicon(); raw=bytes([0x06,0x11,0x20,0xA0])
    assert vqx.decode_ids(raw,lex)==["REQUEST","PEER","RESPOND","GLYPH_ONLY"]
    assert vqx.encode_names(["REQUEST","PEER","RESPOND","GLYPH_ONLY"],lex)==raw


def test_bootstrap_and_flags():
    lex,_,_=load_lexicon(); sample=bytes([0xD3,0xA7,0x5C,0xE1,0x9B,0x02,0x03,0x00,0x06,0x11,0x20,0xA0])
    p=vqx.parse_message(sample,lex,mode="bootstrap")
    assert p["version"]==3 and p["names"]==["REQUEST","PEER","RESPOND","GLYPH_ONLY"]
    expect("VQX_VERSION", lambda: vqx.strip_bootstrap(sample[:6]+bytes([0x99,0])+sample[8:], mode="bootstrap"))
    expect("VQX_FLAGS", lambda: vqx.strip_bootstrap(sample[:7]+bytes([1])+sample[8:], mode="bootstrap"))
    expect("VQX_BEACON", lambda: vqx.strip_bootstrap(bytes([0x06]), mode="bootstrap"))


def test_explicit_compact_avoids_beacon_collision():
    # The family beacon bytes are valid byte values; negotiated compact mode must not heuristically reframe them.
    lex,_,_=load_lexicon(); collision=vqx.BEACON + bytes([0x06])
    p=vqx.strip_bootstrap(collision, mode="compact")
    assert p["mode"]=="compact" and p["payload"]==collision


def test_digest_verification():
    lex,raw,_=load_lexicon(); h=hashlib.sha256(raw).hexdigest()
    parsed=vqx.parse_message(bytes([0x05]),lex,mode="compact",lexicon_bytes=raw,expected_dict_hash=h)
    assert parsed["names"]==["ACK"]
    expect("VQX_DIGEST", lambda: vqx.parse_message(bytes([0x05]),lex,mode="compact",lexicon_bytes=raw,expected_dict_hash="00"*32))
    expect("VQX_DICT_REQUIRED", lambda: vqx.parse_message(bytes([0x05]),lex,mode="compact",expected_dict_hash=h))


def test_macros_and_limits():
    lex,_,_=load_lexicon(); graph=vqx.encode_names(["REQUEST","PEER","RESPOND","GLYPH_ONLY","KEEP_CONTEXT","KEEP_CONSTRAINTS","SEARCH_IF","TOOL_IF","ACCURACY_FIRST","FINAL_ONLY"],lex)
    slot=0xE0; framed=bytes([0xDC,slot,len(graph)])+graph+bytes([0xDD,slot,0xDD,slot,0xDD,slot])
    table=vqx.LocalTable(); expanded=table.expand(framed)
    assert expanded==graph*3 and len(framed)<len(graph)*3 and len(table.table_hash())==64
    expect("VQX_TRUNCATED_MACRO", lambda: vqx.LocalTable().expand(bytes([0xDC])))
    expect("VQX_TRUNCATED_MACRO", lambda: vqx.LocalTable().expand(bytes([0xDC,0xE0,4,1,2])))
    expect("VQX_TRUNCATED_MACRO", lambda: vqx.LocalTable().expand(bytes([0xDD])))
    expect("VQX_LOCAL_UNDEF", lambda: vqx.LocalTable().expand(bytes([0xDD,0xE0])))
    expect("VQX_LOCAL_DIRECT", lambda: vqx.LocalTable().expand(bytes([0xE0])))
    expect("VQX_LOCAL_CONTENT", lambda: vqx.LocalTable().define(0xE0,bytes([0xDD])))
    expect("VQX_EXPANSION_LIMIT", lambda: vqx.LocalTable().expand(framed,max_expanded_size=4))


def test_authority_metadata():
    lex,_,_=load_lexicon(); raw=vqx.encode_names(["REQUEST","PEER","EXECUTE","TOOL"],lex)
    p=vqx.parse_message(raw,lex,mode="compact")
    assert p["authorization_required"]==["EXECUTE"]


def test_size_and_strict_input():
    expect("VQX_FRAME_SIZE", lambda: vqx.wrap_bootstrap(b"12345",max_frame_size=4))
    expect("VQX_NON_PUA", lambda: vqx.pua_to_bytes("hello",strict=True))


def main():
    for fn in [test_roundtrip_pua,test_legacy,test_bootstrap_and_flags,test_explicit_compact_avoids_beacon_collision,test_digest_verification,test_macros_and_limits,test_authority_metadata,test_size_and_strict_input]: fn()
    print("python codec tests ok")

if __name__=="__main__": main()
